### PR TITLE
feat(permit): allowance warning

### LINKED ISSUE #3146: Permit: update allowance warning
# Summary

Update the limit orders allowance warning to account for a permit being executed together with the order.

Right now, if there's no allowance for the given sell token anywhere but as a CoW Hook in the order, the warning will show as if there is no warning (which is true).

Update the warning so it takes into account the permit not yet executed.
It should also identify when a pending permit is no longer valid (nonce updated and allowance not enough for example)

# Background

Potentially move the cached permit storage from "raw" localStorage to Jotai.
See [this comment](https://github.com/cowprotocol/cowswap/pull/3140#discussion_r1336045115) for reference.

### PR DESCRIPTION
**Pull request by alfetopito**
# Summary

Fixes #3146

Update allowance warning check for limit orders

Now the UI will take into account permits still not executed when calculating whether the allowance warning should be displayed:

Before:
image

After:
image

## TODO:

- [x] Move to jotai/updater
- [x] Check periodically
- [ ] Test more use cases
- [ ] Add it to Activity modal (SWAP orders)

  # To Test

1. Make sure there's no allowance
2. Place a LIMIT order for a permittable token
* Allowance warning should NOT be displayed

### CODE REVIEW

File: apps/cowswap-frontend/src/modules/ordersTable/pure/OrdersTableContainer/OrderRow/index.tsx [hunk 1]
```diff
@@ -173,8 +176,17 @@ export function OrderRow({
...
   const showCancellationModal = orderActions.getShowCancellationModal(order)
 
+  const [hasValidPendingPermit, setHasValidPendingPermit] = useState<boolean | undefined>(undefined)
+
+  // TODO: do this properly! Maybe an atom & updater on modules/permit and consume it from there?
+  useEffect(() => {
+    if (checkHasValidPendingPermit && orderParams.hasEnoughAllowance === false) {
+      checkHasValidPendingPermit(order).then(setHasValidPendingPermit)
+    }
+  }, [checkHasValidPendingPermit, order, orderParams.hasEnoughAllowance])
```
Comment (alfetopito) [inline apps/cowswap-frontend/src/modules/ordersTable/pure/OrdersTableContainer/OrderRow/index.tsx:186]: At this point this is checked only once per load.
Working on making this an interval.

File: apps/cowswap-frontend/src/modules/permit/hooks/useCheckHasValidPendingPermit.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+  if (!recoverPermitOwnerPromise) {
+    // The callData doesn't match any known permit type
+    return undefined
+  }
+
+  try {
+    const recoveredOwner = await recoverPermitOwnerPromise
+
+    // Permit is valid when recovered owner matches order owner
+    return recoveredOwner.toLowerCase() === order.owner.toLowerCase()
```
Comment (alfetopito) [inline apps/cowswap-frontend/src/modules/permit/hooks/useCheckHasValidPendingPermit.ts:113]: This is it.
If recovered owner matches order owner, permit is still valid!

File: apps/cowswap-frontend/src/modules/wallet/utils/PermitProviderConnector.ts [hunk 3]
```diff
@@ -42,6 +43,32 @@ export class PermitProviderConnector implements ProviderConnector {
...
+          // @ts-ignore
+          copy[key] = value.toHexString()
+        } else {
+          copy[key] = value
+        }
+      })
+      console.log(`bug--decodeABIParameters`, decodedValues, copy)
+
+      return copy as T
+    }
```
Comment (alfetopito) [inline apps/cowswap-frontend/src/modules/wallet/utils/PermitProviderConnector.ts:70]: Open to suggestions on how to make this better/with no `@ts-ignore`s

File: apps/cowswap-frontend/src/modules/ordersTable/containers/OrdersTableWidget/index.tsx [hunk 4]
```diff
@@ -148,8 +156,23 @@ export function OrdersTableWidget({
...
+        const { hasEnoughAllowance } = getOrderParams(chainId, balancesAndAllowances, item)
+
+        if (hasEnoughAllowance === false) {
+          acc.push(item)
+        }
+      }
+      return acc
+    }, [])
+  }, [balancesAndAllowances, chainId, ordersList.pending])
+
```
Comment (alfetopito) [inline apps/cowswap-frontend/src/modules/ordersTable/containers/OrdersTableWidget/index.tsx:172]: Working on this TODO, not blocking for the review

File: apps/cowswap-frontend/src/modules/permit/const.ts [hunk 5]
```diff
@@ -29,3 +29,5 @@ export const ORDER_TYPE_SUPPORTS_PERMIT: Record<TradeType, boolean> = {
   [TradeType.LIMIT_ORDER]: true,
   [TradeType.ADVANCED_ORDERS]: false,
 }
+
+export const PENDING_ORDER_PERMIT_CHECK_INTERVAL = ms`1min`
```
Comment (alfetopito) [inline apps/cowswap-frontend/src/modules/permit/const.ts:33]: Should we do it more often?

File: apps/cowswap-frontend/src/modules/permit/state/ordersPermitStatusAtom.ts [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import { atom } from 'jotai'
+import { atomWithReset } from 'jotai/utils'
+
+import { OrdersPermitStatus } from '../types'
+
+export const ordersPermitStatusAtom = atomWithReset<OrdersPermitStatus>({})
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/permit/state/ordersPermitStatusAtom.ts:6]: You can use atomWithPartialUpdate() here

File: apps/cowswap-frontend/src/modules/ordersTable/containers/OrdersTableWidget/index.tsx [hunk 7]
```diff
@@ -148,8 +179,11 @@ export function OrdersTableWidget({
 
   useValidatePageUrlParams(orders.length, currentTabId, currentPageNumber)
 
+  const ordersToCheckPendingPermit = useGetOrdersToCheckPendingPermit(ordersList, chainId, balancesAndAllowances)
+
   return (
     <>
+      <PendingPermitUpdater orders={ordersToCheckPendingPermit} />
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/ordersTable/containers/OrdersTableWidget/index.tsx:186]: I'm not sure that `permits` module should depends on `ordersTable`.
Maybe we should move the updater to `App/Updaters.ts`?

File: apps/cowswap-frontend/src/modules/ordersTable/containers/OrdersTableWidget/index.tsx [hunk 8]
```diff
@@ -50,6 +58,28 @@ const ContentWrapper = styled.div`
   width: 100%;
 `
 
+function useGetOrdersToCheckPendingPermit(
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/ordersTable/containers/OrdersTableWidget/index.tsx:61]: Nitpick: to not grow the file size, would be great to move the hook into separate file

File: apps/cowswap-frontend/src/modules/ordersTable/pure/OrdersTableContainer/OrdersTable.tsx [hunk 9]
```diff
@@ -410,6 +412,10 @@ export function OrdersTable({
               if (isParsedOrder(item)) {
                 const order = item
 
+                const orderParams = getOrderParams(chainId, balancesAndAllowances, order)
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/ordersTable/pure/OrdersTableContainer/OrdersTable.tsx:415]: Aha, there is a key of the changes

**Review by shoom3301 — APPROVED** [review]
Works perfectly!
Checked even with different browsers - everything is good.
I've left a couple of comments, but they are about code style, feel free to ignore them.

File: apps/cowswap-frontend/src/modules/permit/state/ordersPermitStatusAtom.ts (hunk 6 above)
Comment (alfetopito) [reply apps/cowswap-frontend/src/modules/permit/state/ordersPermitStatusAtom.ts:6]: Hmmm, actually, I had a different plan in mind when I created the atom.
I no longer need it to be reset, so just a regular atom should do.

Unless you have another idea where `atomWithPartialUpdate` fits better?

Comment (alfetopito) [reply apps/cowswap-frontend/src/modules/permit/state/ordersPermitStatusAtom.ts:6]: Now I see what you mean. That is very handy!

**Review by shoom3301 — APPROVED** [review]

**Comment by alfetopito** [comment]
Merging, please post merge review

**Pull request merged**